From fd607c753bebf155cdac14d02920d69b4a75cd6a Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Ball=C3=B3=20Gy=C3=B6rgy?= <ballogyor@gmail.com>
Date: Wed, 13 Nov 2024 11:54:11 +0100
Subject: [PATCH] Remove URL preview feature

It has security and privacy concerns, and the thumbnailer is not ported to GTK4.

Closes: https://gitlab.gnome.org/GNOME/polari/-/issues/133
Closes: https://gitlab.gnome.org/GNOME/polari/-/issues/159
Closes: https://gitlab.gnome.org/GNOME/polari/-/issues/200
Closes: https://gitlab.gnome.org/GNOME/polari/-/issues/223
---
 src/chatView.js                        |  16 --
 src/meson.build                        |   2 -
 src/org.gnome.Polari.src.gresource.xml |   1 -
 src/thumbnailer.js                     | 241 -------------------------
 src/urlPreview.js                      | 195 --------------------
 5 files changed, 455 deletions(-)
 delete mode 100644 src/thumbnailer.js
 delete mode 100644 src/urlPreview.js

diff --git a/src/chatView.js b/src/chatView.js
index fcb37368..66938b88 100644
--- a/src/chatView.js
+++ b/src/chatView.js
@@ -19,7 +19,6 @@ import {DropTargetIface} from './pasteManager.js';
 import {LogWalker} from './logger.js';
 import {UserPopover} from './userList.js';
 import UserStatusMonitor from './userTracker.js';
-import URLPreview from './urlPreview.js';
 import * as Utils from './utils.js';
 
 export const MAX_NICK_CHARS = 8;
@@ -1293,7 +1292,6 @@ class ChatView extends Gtk.ScrolledWindow {
 
         const channels = Utils.findChannels(text, server);
         const urls = Utils.findUrls(text).concat(channels).sort((u1, u2) => u1.pos - u2.pos);
-        const previews = [];
         let pos = 0;
         for (let i = 0; i < urls.length; i++) {
             const url = urls[i];
@@ -1305,24 +1303,10 @@ class ChatView extends Gtk.ScrolledWindow {
             this._insertWithTags(iter,
                 url.name, tags.concat(this._lookupTag('url'), tag));
 
-            if (GLib.uri_parse_scheme(url.url).startsWith('http'))
-                previews.push(new URLPreview({uri: url.url}));
-
             pos = url.pos + url.name.length;
         }
         this._insertWithTags(iter, text.substr(pos), tags);
 
-        if (previews.length) {
-            this._view.buffer.insert(iter, '\n', -1);
-
-            for (const preview of previews) {
-                this._view.add_child_at_anchor(preview,
-                    this._view.buffer.create_child_anchor(iter));
-            }
-
-            this._view.buffer.insert(iter, '\n', -1);
-        }
-
         if (highlight && message.pendingId) {
             this._pending.set(
                 message.pendingId,
diff --git a/src/meson.build b/src/meson.build
index 0b94b718..b0154f8a 100644
--- a/src/meson.build
+++ b/src/meson.build
@@ -80,5 +80,3 @@ gnome.generate_gir(
 )
 
 subdir('tests')
-
-install_data('thumbnailer.js', install_dir: pkgdatadir)
diff --git a/src/org.gnome.Polari.src.gresource.xml b/src/org.gnome.Polari.src.gresource.xml
index bad7bfe7..8630fb82 100644
--- a/src/org.gnome.Polari.src.gresource.xml
+++ b/src/org.gnome.Polari.src.gresource.xml
@@ -29,7 +29,6 @@ SPDX-FileCopyrightText: 2015 Florian Müllner <fmuellner@gnome.org>
     <file>serverRoomManager.js</file>
     <file>tabCompletion.js</file>
     <file>telepathyClient.js</file>
-    <file>urlPreview.js</file>
     <file>userList.js</file>
     <file>utils.js</file>
     <file>userTracker.js</file>
diff --git a/src/thumbnailer.js b/src/thumbnailer.js
deleted file mode 100644
index 8a9f26fa..00000000
--- a/src/thumbnailer.js
+++ /dev/null
@@ -1,241 +0,0 @@
-// SPDX-FileCopyrightText: 2019 daronion <stefanosdimos.98@gmail.com>
-// SPDX-FileCopyrightText: 2019 Florian Müllner <fmuellner@gnome.org>
-//
-// SPDX-License-Identifier: GPL-2.0-or-later
-
-import Cairo from 'cairo';
-import Gdk from 'gi://Gdk?version=3.0';
-import Gio from 'gi://Gio';
-import GLib from 'gi://GLib';
-import GObject from 'gi://GObject';
-import Gtk from 'gi://Gtk?version=3.0';
-import WebKit2 from 'gi://WebKit2?version=4.1';
-
-import {setConsoleLogDomain} from 'console';
-import {programArgs} from 'system';
-
-Gio._promisify(WebKit2.WebView.prototype, 'get_snapshot');
-Gio._promisify(WebKit2.WebView.prototype, 'evaluate_javascript');
-
-const PREVIEW_WIDTH = 120;
-const PREVIEW_HEIGHT = 90;
-const FALLBACK_ICON_SIZE = 64;
-
-const PreviewWindow = GObject.registerClass(
-class PreviewWindow extends Gtk.Window {
-    static [GObject.properties] = {
-        'uri': GObject.ParamSpec.string(
-            'uri', null, null,
-            GObject.ParamFlags.READWRITE | GObject.ParamFlags.CONSTRUCT_ONLY,
-            null),
-    };
-
-    static [GObject.signals] = {
-        'snapshot-ready': {},
-        'snapshot-failed': {},
-    };
-
-    _snapshot = null;
-
-    constructor(params) {
-        super(params);
-
-        const settings = new WebKit2.Settings({
-            hardware_acceleration_policy: WebKit2.HardwareAccelerationPolicy.NEVER,
-        });
-
-        this._view = new WebKit2.WebView({
-            is_ephemeral: true,
-            visible: true,
-            settings,
-        });
-        this.add(this._view);
-
-        this._view.bind_property('title',
-            this, 'title', GObject.BindingFlags.SYNC_CREATE);
-
-        this._view.connect('authenticate', (view, request) => {
-            request.cancel();
-            return true;
-        });
-
-        this._view.connect('notify::is-loading',
-            this._onLoadingChanged.bind(this));
-        this._view.connect('load-failed', () => this.emit('snapshot-failed'));
-        this._view.load_uri(this.uri);
-
-        GLib.timeout_add_seconds(GLib.PRIORITY_DEFAULT, 60, () => {
-            this._view.stop_loading();
-            return GLib.SOURCE_REMOVE;
-        });
-    }
-
-    _onLoadingChanged() {
-        if (this._view.is_loading)
-            return;
-
-        /* Hopefully wait long enough for a meaningful snapshot,
-           see https://bugs.webkit.org/show_bug.cgi?id=164180 */
-        GLib.timeout_add(GLib.PRIORITY_DEFAULT, 1000, () => {
-            this._createSnapshot();
-            return GLib.SOURCE_REMOVE;
-        });
-    }
-
-    async _createSnapshot() {
-        const getClipOp = this._getImageClip();
-        const snapshotOp = this._view.get_snapshot(
-            WebKit2.SnapshotRegion.VISIBLE,
-            WebKit2.SnapshotOptions.NONE,
-            null);
-        let clip, snapshot;
-
-        try {
-            clip = await getClipOp;
-            snapshot = await snapshotOp;
-        } catch (e) {
-            console.warn(`Failed to create snapshot of ${this.uri}`);
-            console.debug(e);
-            this.emit('snapshot-failed');
-            return;
-        }
-
-        if (clip)
-            this._snapshot = this._createClippedSurface(snapshot, clip);
-        else
-            this._snapshot = snapshot;
-
-        this.emit('snapshot-ready');
-    }
-
-    async _getImageClip() {
-        const script = `
-            const img = document.images[0];
-            document.contentType.startsWith('image')
-                ? [img.x, img.y, img.width, img.height]
-                : null;
-        `;
-
-        let obj = null;
-
-        try {
-            obj = await this._view.evaluate_javascript(script, -1, null, null, null);
-        } catch (e) {
-            console.warn(`Failed to get clip information from ${this.uri}`);
-            console.debug(e);
-        }
-
-        if (!obj || obj.is_null())
-            return null;
-
-        const [x, y, width, height] = obj.object_enumerate_properties()
-            .map(p => obj.object_get_property(p).to_int32());
-
-        if (width === 0 || height === 0)
-            throw new Error('Invalid image clip');
-
-        return {x, y, width, height};
-    }
-
-    _createClippedSurface(source, clip) {
-        const {x, y, width, height} = clip;
-
-        const surface = new Cairo.ImageSurface(Cairo.Format.ARGB32, width, height);
-
-        const cr = new Cairo.Context(surface);
-        cr.setSourceSurface(source, -x, -y);
-        cr.paint();
-        cr.$dispose();
-
-        return surface;
-    }
-
-    getSnapshot() {
-        return this._snapshot;
-    }
-});
-
-class App {
-    constructor(url, filename) {
-        this._uri = url;
-        this._filename = filename;
-    }
-
-    run() {
-        Gtk.init(null);
-
-        const logDomain = 'Polari Thumbnailer';
-        setConsoleLogDomain(logDomain);
-        if (GLib.log_writer_is_journald(2))
-            GLib.setenv('G_MESSAGES_DEBUG', logDomain, false);
-
-        const window = new PreviewWindow({
-            uri: this._uri,
-            default_width: 10 * PREVIEW_WIDTH,
-            default_height: 10 * PREVIEW_HEIGHT,
-        });
-
-        window.realize();
-        window.connect('snapshot-ready', this._onSnapshotReady.bind(this));
-        window.connect('snapshot-failed', this._onSnapshotFailed.bind(this));
-        window.connect('destroy', () => Gtk.main_quit());
-
-        Gtk.main();
-    }
-
-    _onSnapshotReady(window) {
-        const surface = window.getSnapshot();
-        const title = window.title || this._uri;
-        window.destroy();
-
-        if (!surface)
-            return;
-
-        const sourceWidth = surface.getWidth();
-        const sourceHeight = surface.getHeight();
-        const ratio = sourceWidth / sourceHeight;
-
-        let targetWidth, targetHeight;
-        if (ratio >= PREVIEW_WIDTH / PREVIEW_HEIGHT) {
-            targetWidth = Math.min(sourceWidth, PREVIEW_WIDTH);
-            targetHeight = targetWidth / ratio;
-        } else {
-            targetHeight = Math.min(sourceHeight, PREVIEW_HEIGHT);
-            targetWidth = targetHeight * ratio;
-        }
-
-        const target = new Cairo.ImageSurface(Cairo.Format.ARGB32,
-            targetWidth,
-            targetHeight);
-
-        const cr = new Cairo.Context(target);
-        cr.scale(
-            targetWidth / sourceWidth,
-            targetHeight / sourceHeight);
-        cr.setSourceSurface(surface, 0, 0);
-        cr.paint();
-        cr.$dispose();
-
-        const pixbuf = Gdk.pixbuf_get_from_surface(target,
-            0, 0, targetWidth, targetHeight);
-        pixbuf.savev(this._filename, 'png', ['tEXt::Title'], [title]);
-    }
-
-    _onSnapshotFailed(window) {
-        const context = window.get_style_context();
-        context.set_state(Gtk.StateFlags.BACKDROP);
-        const color = context.get_color(context.get_state());
-        window.destroy();
-
-        const [type] = Gio.content_type_guess(this._uri, null);
-        const icon = Gio.content_type_get_symbolic_icon(type);
-        const theme = Gtk.IconTheme.get_default();
-        const info = theme.lookup_by_gicon(icon, FALLBACK_ICON_SIZE, 0);
-        const [pixbuf] = info.load_symbolic(color, null, null, null);
-        pixbuf.savev(this._filename, 'png', [], []);
-    }
-}
-
-const [url, filename] = programArgs;
-const app = new App(url, filename);
-app.run();
diff --git a/src/urlPreview.js b/src/urlPreview.js
deleted file mode 100644
index a4eedc85..00000000
--- a/src/urlPreview.js
+++ /dev/null
@@ -1,195 +0,0 @@
-// SPDX-FileCopyrightText: 2019 daronion <stefanosdimos.98@gmail.com>
-// SPDX-FileCopyrightText: 2019 Florian Müllner <fmuellner@gnome.org>
-// SPDX-FileCopyrightText: 2020 Philip Withnall <withnall@endlessm.com>
-//
-// SPDX-License-Identifier: GPL-2.0-or-later
-
-import GdkPixbuf from 'gi://GdkPixbuf';
-import Gio from 'gi://Gio';
-import GLib from 'gi://GLib';
-import GObject from 'gi://GObject';
-import Gtk from 'gi://Gtk';
-import Pango from 'gi://Pango';
-
-Gio._promisify(Gio._LocalFilePrototype, 'query_info_async');
-Gio._promisify(Gio.Subprocess.prototype, 'wait_async');
-
-class Thumbnailer {
-    static getDefault() {
-        if (!this._singleton)
-            this._singleton = new Thumbnailer();
-        return this._singleton;
-    }
-
-    constructor() {
-        this._urlQueue = [];
-        this._subProc = null;
-        this._thumbnailsDir = `${GLib.get_user_cache_dir()}/polari/thumbnails/`;
-
-        GLib.mkdir_with_parents(this._thumbnailsDir, 0o755);
-    }
-
-    get _hasNetwork() {
-        const monitor = Gio.NetworkMonitor.get_default();
-        return monitor.state_valid && monitor.network_available && !monitor.network_metered;
-    }
-
-    getThumbnail(uri) {
-        return new Promise((resolve, reject) => {
-            const filename = this._generateFilename(uri);
-            const data = {uri, filename, resolve, reject};
-
-            this._processData(data);
-        });
-    }
-
-    async _processData(data) {
-        if (await this._thumbExists(data))
-            this._generationDone(data);
-        else if (!this._hasNetwork)
-            this._generationUnavailable(data);
-        else if (!this._subProc)
-            this._generateThumbnail(data);
-        else
-            this._urlQueue.push(data);
-    }
-
-    _generationDone(data, error = null) {
-        if (error)
-            data.reject(error);
-        else
-            data.resolve(data.filename);
-
-        const nextData = this._urlQueue.shift();
-        if (nextData)
-            this._processData(nextData);
-    }
-
-    _generationUnavailable(data) {
-        this._generationDone(data, new Gio.IOErrorEnum({
-            code: Gio.IOErrorEnum.NETWORK_UNREACHABLE,
-            message: 'Network unreachable',
-        }));
-    }
-
-    async _generateThumbnail(data) {
-        const {filename, uri} = data;
-        this._subProc = Gio.Subprocess.new(
-            ['gjs', '--module', `${pkg.pkgdatadir}/thumbnailer.js`, uri, filename],
-            Gio.SubprocessFlags.NONE);
-        try {
-            await this._subProc.wait_async(null);
-            this._generationDone(data);
-        } catch (e) {
-            this._generationDone(data, e);
-        }
-        this._subProc = null;
-    }
-
-    async _thumbExists(data) {
-        const file = Gio.File.new_for_path(`${data.filename}`);
-        try {
-            await file.query_info_async(Gio.FILE_ATTRIBUTE_STANDARD_TYPE,
-                Gio.FileQueryInfoFlags.NONE, GLib.PRIORITY_DEFAULT, null);
-            return true;
-        } catch {
-            return false;
-        }
-    }
-
-    _generateFilename(url) {
-        const checksum = GLib.Checksum.new(GLib.ChecksumType.MD5);
-        checksum.update(url);
-
-        return `${this._thumbnailsDir}${checksum.get_string()}.png`;
-    }
-}
-
-export default GObject.registerClass(
-class URLPreview extends Gtk.Box {
-    static [GObject.properties] = {
-        'uri': GObject.ParamSpec.string(
-            'uri', null, null,
-            GObject.ParamFlags.READWRITE | GObject.ParamFlags.CONSTRUCT_ONLY,
-            null),
-    };
-
-    constructor(params) {
-        super(params);
-
-        this.set({
-            orientation: Gtk.Orientation.VERTICAL,
-            spacing: 6,
-        });
-
-        this.add_css_class('url-preview');
-        this.add_css_class('background');
-
-        this._imageLoaded = false;
-        this._image = new Gtk.Image();
-        this._image.add_css_class('dim-label');
-        this.append(this._image);
-
-        this._label = new Gtk.Label({
-            halign: Gtk.Align.START,
-            ellipsize: Pango.EllipsizeMode.END,
-        });
-        this._label.add_css_class('dim-label');
-        this.append(this._label);
-
-        this._networkMonitor = Gio.NetworkMonitor.get_default();
-        this._networkChangedId = this._networkMonitor.connect('network-changed',
-            this._maybeLoadImage.bind(this));
-
-        this.connect('destroy', this._onDestroy.bind(this));
-    }
-
-    _onDestroy() {
-        if (this._networkChangedId)
-            this._networkMonitor.disconnect(this._networkChangedId);
-        this._networkChangedId = 0;
-    }
-
-    async _maybeLoadImage() {
-        if (this._imageLoaded || !this.get_mapped())
-            return;
-
-        this._imageLoaded = true;
-        this._image.set({
-            icon_name: 'image-loading-symbolic',
-            pixel_size: 16,
-        });
-        const thumbnailer = Thumbnailer.getDefault();
-
-        let title;
-        try {
-            const filename = await thumbnailer.getThumbnail(this.uri);
-            const pixbuf = GdkPixbuf.Pixbuf.new_from_file(filename);
-
-            title = pixbuf.get_option('tEXt::Title');
-            this._image.set_from_pixbuf(pixbuf);
-            this._image.remove_css_class('dim-label');
-        } catch (e) {
-            if (e.matches(Gio.IOErrorEnum, Gio.IOErrorEnum.NETWORK_UNREACHABLE)) {
-                this._imageLoaded = false;
-            } else {
-                console.info(`Failed to generate thumbnail for ${this.uri}`);
-                console.debug(e);
-            }
-            this._image.set({
-                icon_name: 'image-x-generic-symbolic',
-                pixel_size: 64,
-            });
-        }
-
-        if (title) {
-            this._label.set_label(title);
-            this.tooltip_text = title;
-        }
-    }
-
-    vfunc_map() {
-        super.vfunc_map();
-        this._maybeLoadImage();
-    }
-});
-- 
2.52.0

